src/main/flight/autopilot.c

```diff
@@ -91,14 +190,162 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+                latLong->previousVelocity = velocity;
+                pt1FilterUpdateCutoff(&latLong->accelerationLpf, posHold.pt1Gain);
+                acceleration = pt1FilterApply(&latLong->accelerationLpf, acceleration);
+                float pidA = acceleration * positionPidCoeffs.Kd;
+
+                // limit sum of D and A because otherwise can be too aggressive when starting at speed
+                const float maxDAAngle = 35.0f; // limit in degrees; arbitrary.  20 is a bit too low, allows a lot of overshoot
+                const float pidDA = constrainf(pidD + pidA, -maxDAAngle, maxDAAngle);
+                // note: an angle of more than 35 degrees can still be achieved as P and I grow
+
+                // ** PID Sum **
+                float pidSum = pidP + pidI + pidDA;
+
+                // reset the position target when pidSum crosses zero, typically when velocity is very close to zero, ie craft has stopped
+                // this enhances the smoothness of the transition from stick input back to position hold because there is no sharp change in pidSum
+                if (latLong->isStarting && latLong->pidSum * pidSum < 0.0f) { // pidsum ns has reversed sign
+                    resetPositionControlParams(latLong);
+                    if (i == 0) {
+                        currentTargetLocation.lat = gpsSol.llh.lat;
+                    } else {
+                        currentTargetLocation.lon = gpsSol.llh.lon;
+                    }
+                    latLong->distance = 0.0f;
+                    latLong->isStarting = false;
+                }
+                latLong->pidSum = pidSum;
+
+                // Debugs... distances in cm, angles in degrees * 10, velocities cm/2
+                if (gyroConfig()->gyro_filter_debug_axis == i) {
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 0, lrintf(distanceCm));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 1, lrintf(latLong->distance));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 2, lrintf(latLong->pidSum * 10));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 4, lrintf(pidP * 10));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 5, lrintf(pidI * 10));
+                    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 6, lrintf(pidDA * 10));
+                }
+            }
+        }
+        // ** Rotate pid Sum to quad frame of reference, into pitch and roll **
+        float headingRads = DECIDEGREES_TO_RADIANS(attitude.values.yaw);
+        const float sinHeading = sin_approx(headingRads);
+        const float cosHeading = cos_approx(headingRads);
+
+        posHold.pidSum[AI_ROLL] = -sinHeading * posHold.direction[NORTH_SOUTH].pidSum + cosHeading * posHold.direction[EAST_WEST].pidSum;
+        posHold.pidSum[AI_PITCH] = cosHeading * posHold.direction[NORTH_SOUTH].pidSum + sinHeading * posHold.direction[EAST_WEST].pidSum;
```

Comment:
(see comment about coordinate system. Pitch is in body X axis direction (around Y axis))
```suggestion
        const float sinHeading = sin_approx(-headingRads);
        const float cosHeading = cos_approx(-headingRads);

        posHold.pidSum[AI_PITCH] = cosHeading * posHold.direction[NORTH_SOUTH].pidSum - sinHeading * posHold.direction[EAST_WEST].pidSum;
        posHold.pidSum[AI_ROLL] = sinHeading * posHold.direction[NORTH_SOUTH].pidSum + cosHeading * posHold.direction[EAST_WEST].pidSum;
```

Reply:
@ledvinap @ctzsnooze 
I agree with @ledvinap the sin cos matrix needs to check.
Look at my picture, i agree with @ledvinap.
But for sure, can you write the next information:
- what is positive yaw direction - i think it is CCV in view from top.
- what is pitch direction - i think nose down.
- what is positive roll direction - i think to right side.

Reply:
See comment above.

It was very difficult to get the correct behaviour using the suggested sine and cosine; what I have now works correctly, and I do not want to change it, because it took so much time to get the signs of the distances and the output of cos and sin to provide properly opposing PIDs.

However, I noticed that the PID values are all in the same direction as the distances.  So that for instance a North-South positive distance returned a positive North-South P value.  Usually, we would think of that distance as being an error, so that P would be negative with respect to the distance.  But they are the same sign.

Then when the rotation to quad frame happens, this appears to be reversed.

So there is a 'sign' issue in one place that is corrected in another.

I notice that in the above suggestion, the headings are sign-reversed.  I'd need a matching reverse correction some place earlier on, eg at the point of getting the distances, otherwise the corrections will be wrong.

Reply:
@ctzsnooze Definetelly, there all can not work before with wrong transform matrix! :) And your are right.
But will good understand our axises frame, and angles sign and direction.
I live in other dynamics frame axises standarts, and need some time to use other :)

Reply:
@ctzsnooze : My suggestion only reordered the terms, it should not change the result.
It also 'rotates' result implicitly - you need to rotate around body Y axis (pitch) to get forward (X) motion ... 

> what I have now works correctly, and I do not want to change it, because it took so much time to get the signs of the distances and the output of cos and sin to provide properly opposing PIDs.

But everyone will have to spend similar effort to understand it in future... There are multiple ways to perform rotation. but using 'standard' way makes it much easier to understand.

> However, I noticed that the PID values are all in the same direction as the distances. So that for instance a North-South positive distance returned a positive North-South P value. Usually, we would think of that distance as being an error, so that P would be negative with respect to the distance. But they are the same sign.

P term is usually positive. Error has same direction as action necessary to minimize it.
With heading=0 (north, body X aligned with EF X), you need positive pitch (rotation around body Y axis) to fly north.

> I notice that in the above suggestion, the headings are sign-reversed. I'd need a matching reverse correction some place earlier on, eg at the point of getting the distances, otherwise the corrections will be wrong.

This is problem with BF heading. Normally, for rotations, sin/cos argument is assumed to be anticlockwise (https://en.wikipedia.org/wiki/Rotation_matrix#In_two_dimensions).
You can do the same by changing sign of sin(x) ($$sin( -x) = -sin(x) ; cos (-x) = cos(x)$$), but it is 'common' way to do rotation.

Reply:
@ctzsnooze @ledvinap 
I've checked transforation too: the @ctzsnooze formulas is right  if the headings positive direction is CV.
In my opinion, the yaw angle is more standard for "right hand" defined reference system.
But heading- what is it? I know navigate heading for example, what is CV direction from north. The question is one: What heading angle termin mean in BF?

Reply:
About heading - i've googled some examples CV heading direction...

Reply:
> But heading- what is it? 

Value displayed in OSD (for historic reasons). 
0 - front facing north, 90 - front is facing east

> Personaly, i prefer to compute target xy position in local body system first and compute PID parts after that for real dx dy around drone reference system.

This will be problematic for poshold it quad rotates around Z axis (yaw). I-term will need to be rotated. Using earth frame is a bit easier here (but see discussion about converting PID to ROLL/PITCH).

Reply:
Ok, thank's. 
Oh. Perhaps i've wrote not properly, what i want. I mean compute dx dy first  in local system what is horizontal leveled and rotated at heading angle (no pitch, no roll of course)
Tottaly: the heading direction is CV in BF.

Reply:
> horizontal leveled and rotated at heading angle 

But what happens when heading angle changes rapidly between PID iterations?

Reply:
> Tottaly: the heading direction is CV in BF and @ctzsnooze firmulas are good.

The formulas produce correct results. But they are in nonstandard form, so it takes considerably more time to understand or verify them ( [cos -sin; sin cos] is something you don't have to think about, you just know that is is 2d rotation)

Reply:
> But what happens when heading angle changes rapidly between PID iterations?

The interested question.
But, at first look, we will have new heading position, and new local dx dy orientation. Will get new pitch and roll..

Reply:
Yes, until there is some processing delay (filter or PID I integrator). Then roll and pitch will get mixed (resulting vector will lag behind error)

Reply:
> you just know that is is 2d rotation)

Seems i understand: you want see formulas as standard 2d transform like wiki, for example, formulas.

Reply:
> Yes, until there is some 

@ledvinap We can have very fast coords change during quick heading rotation, more gteate then in earth reference. Thank you!